app/bookmark.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2018 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+class Bookmark(object):
+  """
+  This bookmark object is used as a marker for different places in a
+  text document. Note that because text buffer lines index at 0, all
+  references to rows also assume that 0 is the first line.
+  """
+  def __init__(self, beginRow, endRow, data={}):
+    """
+    Creates a bookmark located at the specified rows. This will automatically
+    cast the passed in rows into integers.
+    Args:
+      beginRow (int): The line that the bookmark starts on (inclusive).
+      endRow (int): The line that the bookmark ends on (inclusive).
+      data (other): This is used to store any information that you would like to
+                    to associate with this bookmark. It is an empty dictionary
+                    by default.
+    """
+    self.__begin = int(beginRow)
+    self.__end = int(endRow)
+    self.data = data
+
+  @property
+  def range(self):
+    return (self.begin, self.end)
+
+  @range.setter
+  def range(self, value):
+    self.__begin, self.__end = int(min(value)), int(max(value))
```

Comment:
While I think it's unlikely to be an issue, is it possible that min/max() may pick a different value than they would if they were ints? e.g. (A, B) where A < B, but int(A) >= int(B).

WDYT of 
```
self.__begin, self.__end = int(value[0]), int(value[1])
if self.__begin > self.__end:
  self.__begin, self.__end = self.__end, self.__begin
```

Reply:
Why not just `sorted(value)` ?

Reply:
Or rather `sorted(int(x) for x in value)` ?

Reply:
> While I think it's unlikely to be an issue, is it possible that min/max() may pick a different value than they would if they were ints? e.g. (A, B) where A < B, but int(A) >= int(B).

Not that I know of? If A < B, and both are of the same sign, then they must both round toward the same direction when casted to an int. If they have differing signs, then int(A) will be <= int(B) since A <= 0 and B >= 0, they'll both round towards 0 (but never switch signs).

Unless there's some weird way of representing floating point numbers like in java that I don't know about..

Reply:
Maybe something silly like
min('4', 5)
5 < '4', but int('4') < int(5)
It seems unlikely, but it also seems possible. Or said another way, with enough time and use of some code, someone is likely to eventually trip over that situation.

Reply:
ooo I see. Okay I just fixed the setter functions. What do you think about it now?

Reply:
In such cases, Python would give:
`TypeError: unorderable types: int() < str()`